arrow-docs/docs/io/README.md

```diff
@@ -12,3 +12,206 @@ Arrow has adopted suspend as system to model monadic computations and offers the
+
+So what guarantees does `IO` bring, and does `suspend` offer the same guarantees?
+
+`IO<A>` always results in `Either<Throwable, A>` whether we run it with `unsafeRunSync` (throws Throwable), `unsafeRunAsync` or `unsafeRunAsyncCancellable`.
+This is done so that any exceptions that occur within the `IO` API can be safely returned to the user, and it can be recovered from at any point in the code.
+Important here is that a `Throwable` that occurs in an async thread is safely captured in the `IO` as well and can be recovered from at any point.
+
+`suspend` always results in `Result<A>` which is isomorphic to `Either<Throwable, A>`, and it can be used to offer the same safety guarantees as `IO`.
+So the `suspend` API can also always return any exception safely to the user, and it can be recovered from at any point in the code.
+In contrast to `IO`, we can only find `startCoroutine` in the standard library and it's isomorphic to `unsafeRunAsync`.
+Instead of `f: (Either<Throwable, A>) -> Unit` you provide `f: (Result<A>) -> Unit` to run the `suspend () -> A` program.
+
+So `suspend () -> A` offers us the exact same guarantees as `IO<A>`.
+
+Both also forces us to go through certain set of methods to execute/run the effectful program.
+
+### Monad transformers vs effect mixing
+
+#### IO<Either<E, A>> vs suspend () -> Either<E, A>
+
+When writing functional code style we often want to express our domain errors as clearly as possible, a popular pattern is to return `Either<DomainError, SuccessValue>`.
+Let's assume following domain, and compare two snippets one using `IO<Either<E, A>>` and another `suspend () -> Either<E, A>`.
+
+```kotlin:ank
+import arrow.core.Either
+import arrow.core.Left
+import arrow.core.Right
+
+inline class Id(val id: Long)
+object PersistenceError
+
+data class User(val email: String, val name: String)
+data class ProcessedUser(val id: Id, val email: String, val name: String)
+
+suspend fun fetchUser(): Either<PersistenceError, User> =
+  Right(User("simon@arrow-kt.io", "Simon"))
+
+suspend fun User.process(): Either<PersistenceError, ProcessedUser> =
+  if (email.contains(Regex("^(.+)@(.+)$"))) Right(ProcessedUser(Id(1), email, name))
+  else Left(PersistenceError)
+```
+
+###### IO<Either<E, A>>
+
+ ```kotlin:ank
+import arrow.fx.*
+
+fun ioProgram(): IO<Either<PersistenceError, ProcessedUser>> =
+  IO.fx {
+    val res = !IO.effect { fetchUser() }
+
+    !res.fold({ error ->
+      IO.just(Either.Left(error))
+    }, { user ->
+      IO.effect { user.process() }
+    })
+  }
+
+// Or unwrapped in `suspend`
+suspend suspendedIOProgram(): Either<PersistenceError, ProcessedUser> =
+  ioProgram().suspended()
+ ```
+
+##### suspend () -> Either<E, A>
+
+```kotlin:ank
+import arrow.core.computations.either
+
+suspend fun suspendProgram(): Either<PersistenceError, ProcessedUser> =
+  either {
+    val user = !fetchUser()
+    val processed = !user.process()
+    processed
+  }
+```
+
```

Comment:
```suggestion
The above two examples demonstrate how much simpler `suspend` is over its `IO` counterpart and how the `either` computation block allows us to bind values of `Either` to extract their right side all while inside `suspend`.

Arrow Fx allows to intermix effects in suspension. What otherwise would have required the `EitherT` transformer over `IO` now it can just be expressed by wrapping in `either` instead
```